utils/llm.py

```diff
@@ -2,130 +2,98 @@
-        return 4_096
-    model = get_model(model)
-    prompt_size, _ = SUPPORTED_MODELS[model]
-    return prompt_size
+    model = model_from_alias(model)
+    return litellm.model_cost[model]["max_input_tokens"]
 
 
 def token_cost(model: str) -> tuple[float, float]:
-    if model == "text-embedding-ada-002":
-        return 2e-9, 2e-9
-    if model == "gpt-3.5-turbo-instruct":
-        return 1.5e-6, 2e-6
-    model = get_model(model)
-    _, cost_info = SUPPORTED_MODELS[model]
-    return cost_info
-
-
-def response_cost(response: openai.ChatCompletion) -> float:
-    input_cost, output_cost = token_cost(response.model)
-    usage = response.usage
-    return usage.prompt_tokens * input_cost + usage.completion_tokens * output_cost
+    model = model_from_alias(model)
+    input_cost = litellm.model_cost[model]["input_cost_per_token"]
+    output_cost = litellm.model_cost[model]["output_cost_per_token"]
+    return input_cost, output_cost
 
 
 def set_api_key():
-    if openai.api_key is not None:
-        return
-    openai.api_key = os.getenv("OPENAI_API_KEY")
+    try:
+        if litellm.openai_key is None:
+            litellm.openai_key = os.getenv("OPENAI_API_KEY")
+    except:
+        pass
+
+    try:
+        if litellm.anthropic_key is None:
+            litellm.anthropic_key = os.getenv("ANTHROPIC_API_KEY")
+    except:
+        pass
 
 
 def ensure_context_len(
     context: LLMContext,
-    model: Optional[str] = None,
+    model: str = CHEAPEST_TOKENISER,
     max_len: Optional[int] = None,
     response_len: int = 0,
-    system_message: bool = True,
 ) -> tuple[LLMContext, int]:
+    model = model_from_alias(model)
     max_len = max_len or get_max_prompt_size(model)
     messages = list()
+    context_tokens = litellm.token_counter(model, messages=context[:1])
 
-    if system_message:
-        sys_idx = 1
-    else:
-        sys_idx = 0
-
-    context_tokens = context_token_len(context[:sys_idx], model=model)
-
-    for message in reversed(context[sys_idx:]):
-        message_tokens = context_token_len([message], model=model)
+    for message in reversed(context[1:]):
+        message_tokens = litellm.token_counter(model, text=message["content"])
         if context_tokens + message_tokens + response_len > max_len:
             break
         messages.append(message)
         context_tokens += message_tokens
     messages.reverse()
-    context = context[:sys_idx] + messages
+    context = context[:1] + messages
     # assert len(context) > 1, f"There are messages missing in the context:\n\n{context}"
     return context, context_tokens
 
 
 def ask_llm(
     context: LLMContext,
-    model: Optional[str] = None,
+    model: str,
     temperature: float = 1,
     context_length: int = None,
     cost_callback: Callable[[float], None] = None,
     timeout: float = 300,
-    max_tokens: Optional[int] = None,
+    max_response_tokens: Optional[int] = None,
```

Comment:
Not as it currently stands, because `ensure_context_len` uses it and expects it to be an integer. What I can do is just get the `max_tokens` for the model right after getting the model name.

Reply:
Ah I understand now. Yeah, you can do that and also use the typing
```python
max_response_tokens: int = None
```
which should make clear that the value is expected to be filled at some point, and the null value is just there to ease checks.